### ISSUE TITLE
Can't use multiple XLA devices when using PyTorch XLA

### ISSUE THREAD

**Issue by humzaiqbal** [UCSC-VLAA/CLIPA#6]
## OS 
TPU version: v2-8
OS: Ubuntu 20.04.4 LTS
torch version : 2.0.0+cu117
torch_xla version: 2.0
python version: 3.8.10

## Issue 
I want to run distributed training using a TPU v2-8 which should have 8 XLA devices. To do so I ran a script using the following 
```
python3 /home/ubuntu/CLIPA/clipa_torch/launch_xla.py --num-devices 8 training.main \
    --train-data /path/to/train/data \
    --val-data /path/to/val/data \
    --batch-size 80 \
    --epochs 20 \
    --pretrained /path/to/clipa_vit_b16_i50_t16.pt \
    --lr 5e-05 \
    --model ViT-B-16-CL16 \
    --precision 'fp32' \
    --train-num-samples 2515000 \
    --warmup 1000 \
```
When I do so the code reaches [this assert](https://github.com/UCSC-VLAA/CLIPA/blob/867ddebaaf0dec47bfd7baaf97b79213eac62ab1/clipa_torch/training/main.py#L103) and breaks. Is using more than 1 XLA device not supported with this code?

For what its worth when I comment out the assert I get this error 
```
2023-08-23 17:52:13.861705: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
concurrent.futures.process._RemoteTraceback:
"""
Traceback (most recent call last):
  File "/usr/lib/python3.8/concurrent/futures/process.py", line 239, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/usr/lib/python3.8/concurrent/futures/process.py", line 198, in _process_chunk
    return [fn(*args) for args in chunk]
  File "/usr/lib/python3.8/concurrent/futures/process.py", line 198, in <listcomp>
    return [fn(*args) for args in chunk]
  File "/usr/local/lib/python3.8/dist-packages/torch_xla/experimental/pjrt.py", line 92, in wrapper
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/torch_xla/experimental/pjrt.py", line 245, in _run_thread_per_device
    replica_results = list(
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 619, in result_iterator
    yield fs.pop().result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 444, in result
    return self.__get_result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.8/dist-packages/torch_xla/experimental/pjrt.py", line 238, in _thread_fn
    return fn()
  File "/usr/local/lib/python3.8/dist-packages/torch_xla/experimental/pjrt.py", line 341, in __call__
    self.fn(global_ordinal(), *self.args, **self.kwargs)
  File "/home/ubuntu/CLIPA/clipa_torch/training/main.py", line 506, in _mp_entry
    main(sys.argv[1:])
  File "/home/ubuntu/CLIPA/clipa_torch/training/main.py", line 104, in main
    date_str = broadcast_object(args, date_str)
  File "/home/ubuntu/CLIPA/clipa_torch/training/distributed.py", line 139, in broadcast_object
    dist.broadcast_object_list(objects, src=src)
  File "/usr/local/lib/python3.8/dist-packages/torch/distributed/distributed_c10d.py", line 1436, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/torch/distributed/distributed_c10d.py", line 2230, in broadcast_object_list
    current_device = device or _get_pg_device(group)
  File "/usr/local/lib/python3.8/dist-packages/torch/distributed/distributed_c10d.py", line 419, in _get_pg_device
    if _check_for_nccl_backend(group):
  File "/usr/local/lib/python3.8/dist-packages/torch/distributed/distributed_c10d.py", line 1959, in _check_for_nccl_backend
    pg = group or _get_default_group()
  File "/usr/local/lib/python3.8/dist-packages/torch/distributed/distributed_c10d.py", line 697, in _get_default_group
    raise RuntimeError(
RuntimeError: Default process group has not been initialized, please make sure to call init_process_group.
```

**Comment by humzaiqbal** [comment]
I tried modifying the code to use distributed PyTorch XLA backed by PJRT (note: this may require torch / torch_xla to be >= 2.0) and I get this now

```
 File "/usr/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch_xla/experimental/pjrt.py", line 238, in _thread_fn
    return fn()
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch_xla/experimental/pjrt.py", line 341, in __call__
    self.fn(global_ordinal(), *self.args, **self.kwargs)
  File "/home/ubuntu/CLIPA/clipa_torch/training/main.py", line 506, in _mp_entry
    main(sys.argv[1:])
  File "/home/ubuntu/CLIPA/clipa_torch/training/main.py", line 104, in main
    date_str = broadcast_object(args, date_str)
  File "/home/ubuntu/CLIPA/clipa_torch/training/distributed.py", line 144, in broadcast_object
    dist.broadcast_object_list(objects, src=src)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 1451, in wrapper
    return func(*args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 2255, in broadcast_object_list
    broadcast(object_sizes_tensor, src=src, group=group)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 1451, in wrapper
    return func(*args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 1566, in broadcast
    work = default_pg.broadcast([tensor], opts)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch_xla/distributed/xla_backend.py", line 94, in broadcast
    xm.collective_broadcast([root_tensor],
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch_xla/core/xla_model.py", line 798, in collective_broadcast
    xscale = send_cpu_data_to_device(scale, tensor.device)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch_xla/core/xla_model.py", line 1092, in send_cpu_data_to_device
    return ToXlaTensorArena(convert_fn, select_fn).transform(data)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch_xla/core/xla_model.py", line 431, in transform
    self._convert()
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch_xla/core/xla_model.py", line 403, in _convert
    self._converted_tensors = self._convert_fn(self._tensors)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch_xla/core/xla_model.py", line 1087, in convert_fn
    return torch_xla._XLAC._xla_tensors_from_aten(tensors, devices)
RuntimeError: /pytorch/xla/torch_xla/csrc/aten_xla_bridge.cpp:276 : Check failed: device.type() == at::kXLA (cpu vs. xla)
*** Begin stack trace ***
        tsl::CurrentStackTrace()
        torch_xla::bridge::AtenDeviceToXlaDevice(c10::Device const&)



        PyCFunction_Call
        _PyObject_MakeTpCall
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall
        _PyE
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        PyObject_Call
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall

        PyObject_Call

        _PyObject_MakeTpCall
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        PyObject_Call
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall
        PyObject_Call
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall

        PyObject_Call



        clone
*** End stack trace ***
cpu
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/ubuntu/CLIPA/clipa_torch/launch_xla.py", line 64, in <module>
    main()
  File "/home/ubuntu/CLIPA/clipa_torch/launch_xla.py", line 60, in main
    xmp.spawn(mod._mp_entry, args=(), nprocs=args.num_devices)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 386, in spawn
    return pjrt.spawn(fn, nprocs, start_method, args)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch_xla/experimental/pjrt.py", line 365, in spawn
    _run_multiprocess(spawn_fn, start_method=start_method)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch_xla/experimental/pjrt.py", line 92, in wrapper
    return fn(*args, **kwargs)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch_xla/experimental/pjrt.py", line 322, in _run_multiprocess
    replica_results = list(
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch_xla/experimental/pjrt.py", line 323, in <genexpr>
    itertools.chain.from_iterable(
  File "/usr/lib/python3.8/concurrent/futures/process.py", line 484, in _chain_from_iterable_of_lists
    for element in iterable:
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 619, in result_iterator
    yield fs.pop().result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 444, in result
    return self.__get_result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
RuntimeError: /pytorch/xla/torch_xla/csrc/aten_xla_bridge.cpp:276 : Check failed: device.type() == at::kXLA (cpu vs. xla)
*** Begin stack trace ***
        tsl::CurrentStackTrace()
        torch_xla::bridge::AtenDeviceToXlaDevice(c10::Device const&)



        PyCFunction_Call
        _PyObject_MakeTpCall
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        PyObject_Call
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        PyObject_Call
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        PyObject_Call
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall

        PyObject_Call

        _PyObject_MakeTpCall
        _PyEval_EvalFrameDefault
        _PyEval_EvalCodeWithName
        _PyFunction_Vectorcall
        PyObject_Call
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall
        PyObject_Call
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall
        _PyEval_EvalFrameDefault
        _PyFunction_Vectorcall

        PyObject_Call



        clone
*** End stack trace ***
cpu
```

Seems like the actual error is this part 

```
RuntimeError: /pytorch/xla/torch_xla/csrc/aten_xla_bridge.cpp:276 : Check failed: device.type() == at::kXLA (cpu vs. xla)
*** Begin stack trace ***
```
seems there is some device error where something is on the CPU and something on the TPU but hard to parse out what exactly

**Comment by zw615** [comment]
Hi there! 

This issue is because you didn't specify `args.name` in the command line, and thus the `broadcast_object` function is called (see line 104 in main.py). So far this function is only supported when the `init_process_group` function is called. In other words, it is only supported on GPU. A simple fix is to specify the name of the experiment in the command line.

Please do not hesitate to let me know if you have any more question.
